src/controllers/upsert.js

```diff
@@ -1,28 +1,56 @@
 const core = require('@actions/core');
+const github = require('@actions/github');
 
+const circleci = require('../circleci');
 const comments = require('../comments');
 const git = require('../git');
 const heroku = require('../heroku');
 
-async function upsertController(params) {
-  const { pipelineName, pipelineId, logDrainUrl, appName, refName } = params;
+async function deployDocker(appName, nodeEnv) {
+  const owner = github.context.payload.repository.owner.login;
+  const repo = github.context.payload.repository.name;
+  const branch = github.context.payload.pull_request.head.ref;
+  const pipeline = await circleci.startDockerBuild(owner, repo, branch, appName, nodeEnv);
+
+  core.info(`  - Kicked off CircleCI pipeline #${pipeline.number}. Waiting for pipeline to finish;`);
+  core.info('    this may take some time. Watch the build progress here:');
+  core.info(`    https://app.circleci.com/pipelines/github/${owner}/${repo}/${pipeline.number}`);
+
+  const result = await circleci.waitForPipelineFinish(pipeline.id);
+  if (result.status === 'success') {
+    core.info('    Build and deploy pipeline finished successfully!');
+  } else {
+    throw new Error(`CircleCI pipeline #${pipeline.number} finished with status "${result.status}".`);
+  }
+}
 
-  if (!git.refExists(refName)) {
-    throw new Error(`Ref "${refName}" does not exist.`);
+async function deployHeroku(appName, refName) {
+  const credentials = heroku.getCredentials();
+  const pushResult = git.push(credentials, appName, refName);
+  if (pushResult.status !== 0) {
+    throw new Error(`Created Heroku app "${appName}" but ran into errors deploying.`);
   }
+}
+
+async function upsertController(params) {
+  const { pipelineName, pipelineId, logDrainUrl, appName, refName, useDocker } = params;
+
   const sha = git.shaForRef(refName); // can't use github.context.sha because we want to exclude merge commits
   const message = git.messageForRef(refName);
   const configVars = await heroku.getPipelineVars(pipelineId);
 
   let stepCount = 4;
```

Comment:
I'm thinking about refactoring all this "step" stuff to something with a lot more structure... something where each step is encapsulated in an object, so the controller would look a lot more like this:

```javascript
const steps = [];

const createApp = new CreateAppStep();
if (createApp.isNeeded()) {
  steps.push(createApp);
}

if (useDocker) {
   steps.push(new DockerDeployStep());
} else {
   steps.push(new HerokuDeployStep());
}

for (let i=0; i<steps.length; i++) {
  steps[i].run();
}
```

But I wanted to keep that outside the scope of this PR, otherwise it was going to be too much.

Reply:
OMG I have a refactor that's _amazing_. 🎉 Will send this as a separate PR tomorrow. Here's an example step:

```javascript
class CreateAppStep {
  constructor(params) {
    this.params = params;
    this.shouldRun = undefined;
    this.title = `Creating Heroku app "${params.appName}"`;
  }

  async checkPrereqs() {
    this.shouldRun = !(await heroku.appExists(this.params.appName));
  }

  async run() {
    return heroku.createApp(this.params.appName, this.params.pipelineId);
  }
}
```